corehq/apps/users/bulk_download.py

```diff
@@ -43,84 +47,136 @@ def build_data_headers(keys, header_prefix='data'):
+    try:
+        location_codes.append(location_cache.get(user.location_id))
+    except SQLLocation.DoesNotExist:
+        pass
+    for location_id in user.assigned_location_ids:
+        # skip if primary location_id, as it is already added to the start of list above
+        if location_id != user.location_id:
+            try:
+                location_codes.append(location_cache.get(location_id))
+            except SQLLocation.DoesNotExist:
+                pass
+
+    def _format_date(date):
+        return date.strftime('%Y-%m-%d %H:%M:%S') if date else ''
+
+    return {
+        'data': model_data,
+        'uncategorized_data': uncategorized_data,
+        'group': group_names,
+        'name': user.full_name,
+        'password': "********",  # dummy display string for passwords
+        'phone-number': user.phone_number,
+        'email': user.email,
+        'username': user.raw_username,
+        'language': user.language,
+        'user_id': user._id,
+        'is_active': str(user.is_active),
+        'User IMEIs (read only)': get_devices(user),
+        'location_code': location_codes,
+        'role': role.name if role else '',
+        'user_profile': profile.name if profile else '',
+        'registered_on (read only)': _format_date(user.created_on),
+        'last_submission (read only)': _format_date(activity.last_submission_for_user.submission_date),
+        'last_sync (read only)': activity.last_sync_for_user.sync_date,
+    }
+
+
+def make_web_user_dict(user, domain):
+    user = CouchUser.wrap_correctly(user['doc'])
+    domain_membership = user.get_domain_membership(domain)
+    role = UserRole.get(domain_membership.role_id).name
+    return {
+        'first_name': user.first_name,
+        'last_name': user.last_name,
+        'email': user.email,
+        'username': user.username,
+        'status': 'Active User',
+        'role': role,
+        'last_access_date (read only)': domain_membership.last_accessed,
+        'last_login (read only)': user.last_login,
+        'remove': '',
+    }
+
+
+def get_role_name(role):
+    if role:
+        if role == 'admin':
+            return role
+        else:
+            role_id = role[len('user-role:'):]
+            try:
+                return UserRole.get(role_id).name
+            except ResourceNotFound:
+                return ugettext('Unknown Role')
+    else:
+        return None
+
+
+def make_invited_web_user_dict(invite):
+    return {
+        'first_name': 'N/A',
+        'last_name': 'N/A',
+        'email': invite.email,
+        'username': invite.email,
+        'status': 'Invited',
+        'role': get_role_name(invite.role),
+        'last_access_date (read only)': 'N/A',
+        'last_login (read only)': 'N/A',
+        'remove': '',
+    }
+
+
+def get_user_rows(user_dicts, user_headers):
+    for user_dict in user_dicts:
+        row = dict(flatten_json(user_dict))
+        yield [row.get(header) or '' for header in user_headers]
```

Comment:
Nitpick: `row.get(header, '')` is a more common way to write `row.get(header) or ''`

Reply:
There's also a difference depending on what falsy values are possible in this context.  If a user field is `None`, I'm guessing it should be represented by `''`, but if it's `False` or `0`, probably not.